### PR TITLE
Fix 180 uap/uapaot tests in System.IO.MemoryMappedFiles.Tests

### PR DESCRIPTION
**Pull request by ViktorHofer**
MemoryMappedFileAccess.ReadExecute or MemoryMappedFileAccess.ReadWriteExecute isn't permitted inside an AppContainer.

Fixes 180 tests here:
https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fuwp~2F/build/20170515.01/workItem/System.IO.MemoryMappedFiles.Tests

### CODE REVIEW

File: src/Common/tests/System/AssertExtensions.cs [hunk 1]
```diff
@@ -104,5 +104,18 @@ public static void ThrowsAny(Type firstExceptionType, Type secondExceptionType,
         {
            ThrowsAny(typeof(TFirstExceptionType), typeof(TSecondExceptionType), action);
         }
+
+        public static void ThrowsIf<T>(Action action, bool condition)
```
Comment (stephentoub) [inline src/Common/tests/System/AssertExtensions.cs]: I suggest the arguments be swapped here to have the condition be first.  That better aligns with APIs like Debug.WriteLineIf.

File: src/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateNew.Tests.cs [hunk 2]
```diff
@@ -152,10 +152,15 @@ public void ValidMapNames_Windows(string name)
             {
                 ValidateMemoryMappedFile(mmf, 4096, MemoryMappedFileAccess.Read);
             }
-            using (MemoryMappedFile mmf = MemoryMappedFile.CreateNew(name, 4096, MemoryMappedFileAccess.ReadWriteExecute, MemoryMappedFileOptions.DelayAllocatePages, HandleInheritability.Inheritable))
+
+            // MemoryMappedFileAccess.ReadExecute or MemoryMappedFileAccess.ReadWriteExecute isn't permitted inside an AppContainer
```
Comment (stephentoub) [inline src/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateNew.Tests.cs]: Basically just Execute permissions aren't allowed, right?

File: src/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateNew.Tests.cs [hunk 3]
```diff
@@ -195,14 +200,23 @@ public void ReusingNames_Windows(string name)
...
+
+            AssertExtensions.ThrowsIf<UnauthorizedAccessException>(() =>
             {
-                ValidateMemoryMappedFile(mmf, capacity, access, inheritability);
-            }
+                using (MemoryMappedFile mmf = MemoryMappedFile.CreateNew(mapName, capacity, access, options, inheritability))
+                {
+                    ValidateMemoryMappedFile(mmf, capacity, access, inheritability);
+                }
+            }, PlatformDetection.IsWinRT && (access == MemoryMappedFileAccess.ReadExecute|| access == MemoryMappedFileAccess.ReadWriteExecute));
```
Comment (stephentoub) [inline src/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateNew.Tests.cs]: Nit: space before the `||`

File: src/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateNew.Tests.cs [hunk 4]
```diff
@@ -195,14 +200,23 @@ public void ReusingNames_Windows(string name)
             {
                 ValidateMemoryMappedFile(mmf, capacity);
             }
-            using (MemoryMappedFile mmf = MemoryMappedFile.CreateNew(mapName, capacity, access))
+
+            // MemoryMappedFileAccess.ReadExecute or MemoryMappedFileAccess.ReadWriteExecute isn't permitted inside an AppContainer
```
Comment (joperezr) [inline src/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateNew.Tests.cs]: NIT: should we just remove this comment and instead have this in the commit message? that way if people want to figure out why we added this they can go in the history.

**Comment by joperezr** [comment]
Small nit and question, but LGTM otherwise.

**Comment by danmosemsft** [comment]
@stephentoub @ViktorHofer can you think of any reason why it shouldn't be allowed? If not, we should open a bug against Windows (now, while they still have time). I had assumed AppContainer was merely tighter ACLs (effectively ,by being lower integrity)

In general I'm thinking we need to prioritize the F5 runs highest, in order to find any other unexpected limitations.

**Comment by ViktorHofer** [comment]
My guess is that running shared executable code from memory isn't allowed. https://msdn.microsoft.com/en-us/library/system.io.memorymappedfiles.memorymappedfileaccess(v=vs.110).aspx

ReadWriteExecute
Read and write access to the file that can can store and run executable code.

**Comment by ViktorHofer** [comment]
> In general I'm thinking we need to prioritize the F5 runs highest, in order to find any other unexpected limitations.

I agree!

File: src/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateNew.Tests.cs (hunk 4 above)
Comment (ViktorHofer) [inline src/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateNew.Tests.cs]: Yes, will do

File: src/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateNew.Tests.cs (hunk 2 above)
Comment (ViktorHofer) [inline src/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateNew.Tests.cs]: Correct

**Comment by ViktorHofer** [comment]
> Dumb question: is PlatformDetection.IsWinRT set for the F5 uap tests? I assume so, and if it is, did you try running the F5 uap tests as well to make sure that this changes won't break tests in those legs?

@danmosemsft told me that IsWinRT means that you are inside an AppContainer. This *should* apply fort both UAP and UAPAOT. Yeah I tested it :)

**Comment by joperezr** [comment]
> @danmosemsft told me that IsWinRT means that you are inside an AppContainer. This should apply fort both UAP and UAPAOT. Yeah I tested it :)

FYI, AFAIK ILC tests are NOT yet running on an AppContainer, so not really sure if that is how its set given that you have verified that this change will make them pass.

**Comment by ViktorHofer** [comment]
> FYI, AFAIK ILC tests are NOT yet running on an AppContainer, so not really sure if that is how its set given that you have verified that this change will make them pass.

I know. What I meant is, that UAPAOT definitely should run inside an AppContainer. I verified the behavior on UAP.

**Comment by danmosemsft** [comment]
I'm OK with disabling these for now but @ViktorHofer is going to help start an email thread with Windows to see whether they should fix this or not.

**Pull request merged**